src/driver-holtek/holtek8-shared.c

```diff
@@ -0,0 +1,1291 @@
+					delay = 0;
+				}
+				else if (delay != 0) {
+					if (delay / delay_base_ms < 128) {
+						raw_delay = delay / delay_base_ms;
+						if (raw_delay == 0)
+							raw_delay = 1;
+
+						raw[event_i-1].delay = raw_delay;
+					}
+					else {
+						raw_delay = delay / (2 * delay_base_ms);
+						raw[event_i++].key = HOLTEK8_MACRO_CMD_WAIT;
+						set_unaligned_be_u16((uint8_t*) &raw[event_i++], raw_delay);
+					}
+					delay = 0;
+				}
+
+				raw[event_i].release = ratbag_ev->type == RATBAG_MACRO_EVENT_KEY_RELEASED;
+				raw[event_i].delay = 1;
+
+				switch (ratbag_ev->event.key) {
+					case BTN_LEFT:   key = HOLTEK8_BUTTON_MOUSE_LEFT; break;
+					case BTN_RIGHT:  key = HOLTEK8_BUTTON_MOUSE_RIGHT; break;
+					case BTN_MIDDLE: key = HOLTEK8_BUTTON_MOUSE_MIDDLE; break;
+					case BTN_SIDE:   key = HOLTEK8_BUTTON_MOUSE_MB4; break;
+					case BTN_EXTRA:  key = HOLTEK8_BUTTON_MOUSE_MB5; break;
+					default:
+						key = ratbag_hidraw_get_keyboard_usage_from_keycode(device, ratbag_ev->event.key);
+				}
+				if (key == 0)
+					return -EINVAL;
+
+				raw[event_i].key = key;
+				event_i += 1;
+				break;
+			case RATBAG_MACRO_EVENT_WAIT:
+				delay += ratbag_ev->event.timeout;
+				break;
+			case RATBAG_MACRO_EVENT_NONE:
+				return event_i;
+			default:
+				return -EINVAL;
+		}
+	}
+
+	return event_i;
+}
+
+/*
+ * Converts raw device button data and writes to ratbag button.
+ * If the data->type is macro, reads a decoded macro from the device
+ * memory and writes to ratbag button.
+ *
+ * @return 0 on success or a negative errno.
+ */
+int
+holtek8_button_from_data(struct ratbag_button *button, const struct holtek8_button_data *data)
+{
+	struct ratbag_device *device = button->profile->device;
+	const struct ratbag_button_action *action;
+	int rc;
+
+	action = holtek8_button_map_raw_to_action(data);
+
+	if (action != NULL) {
+		button->action = *action;
+		return 0;
+	}
+
+	button->action.type = RATBAG_BUTTON_ACTION_TYPE_UNKNOWN;
+
+	switch (data->type) {
+		case HOLTEK8_BUTTON_TYPE_KEYBOARD: {
+			const unsigned int modifiers = data->keyboard.modifiers;
+			const unsigned int key1 = ratbag_hidraw_get_keycode_from_keyboard_usage(device, data->keyboard.hid_key);
+			const unsigned int key2 = ratbag_hidraw_get_keycode_from_keyboard_usage(device, data->keyboard.hid_key2);
+
+			holtek8_button_macro_new_from_keycodes(button, key1, key2, modifiers);
+			break;
+		}
+		case HOLTEK8_BUTTON_TYPE_MEDIA: {
+			const uint16_t hid_code_cc = get_unaligned_le_u16(data->media.hid_key);
+			const unsigned int key = ratbag_hidraw_get_keycode_from_consumer_usage(device, hid_code_cc);
+
+			holtek8_button_macro_new_from_keycodes(button, key, 0, 0);
+			break;
+		}
+		case HOLTEK8_BUTTON_TYPE_MACRO: {
+			struct holtek8_macro_event macro_events[HOLTEK8_MAX_MACRO_EVENTS] = {0};
+
+			rc = holtek8_read_macro_data(device, macro_events, data->macro.index);
+			if (rc == -EOVERFLOW)
+				return 0;
```

Comment:
I don't really understand this specific check.
HOLTEK8_MAX_MACRO_EVENTS and MAX_MACRO_EVENTS are both 256, and they store more or less the same information, so in the end:
1) I would expect this check to never be true.
2) Even if would be, this would silently ignore the error. I would at least expect a warning like 'there are more macro events than libratbag can handle, showing empty macro instead'.

Reply:
This check is after reading from memory and before parsing. It can overflow if the macro which already is in device memory has too many jumps (or even infinite loops), although I'm not sure if official software can generate macro this big, but we never know. Added a log here.